### ISSUE TITLE
RFC: changing syncoid from explicit rollback to receive -F

### ISSUE THREAD

**Issue by jimsalterjrs** [jimsalterjrs/sanoid#735]
Long story short: I wrote an article on ZFS permission delegation recently, and as a result I realized that using syncoid requires more privileges than replication alone does.

Currently, syncoid explicitly rolls back the target to the most recent common snapshot prior to replication. So if we have @1-@4 on the target, and we have @1-@6 on the source, syncoid performs zfs rollback -r pool/ds1@4 on the target before attempting replication.

Here's the problem: this requires both receive and potentially destroy privileges be granted to the receiving user. On the other hand, if we receive using the -F flag, we don't need either privilege... even when the receive operation actually does end up trashing ephemeral snapshots or data on the target.

For security's sake, I'd therefore like to shift syncoid from explicit rollbacks to simply receiving with -F. I can't think of any scenarios where that would be a problem.

Comments?

**Comment by redmop** [comment]
How will this affect the @syncoid snapshots? Don't they need destroying when they are finished being used?

**Comment by jimsalterjrs** [comment]
Yes, if you want to use sync snapshots, you need `zfs allow username destroy poolname/datasetname` on both sides.

You can just use --no-sync-snap and avoid that issue, though, if you have something like Sanoid automating snapshots on the source.

**Comment by redmop** [comment]
Just making sure that is considered. I normally use --no-sync-snap.

**Comment by jimsalterjrs** [comment]
@phreaker0 is it time to package up a new release? removing rollbacks is significant because it means fewer privs must be delegated to non-privileged users for more secure replication, and that local replication fail bug when using -s strikes me as a big deal; honestly I'm surprised we never encountered it until today.

**Comment by prewashed** [comment]
Unfortunately, this change breaks my use. I have a number of clients that syncoid to a server. That server gets backed up onto a number of external disks. 

A) syncoid several clients to a remote backup pool (S)
    -r --no-rollback

B) syncoid S to several local removable backup disks (ext[0-N]) using:
    -r --identifier=extn --force-delete

A and B occur in no particular order, e.g.

  1.  A syncoid client1 to S
  2.  A syncoid client3 to S
  3.  B syncoid S to ext3
  4.  A syncoid client 1 to S
  5.  B syncoid S to ext1
  6.  A syncoid client 2 to S
  7.  B syncoid S to ext3

Step 3's syncoid to external ext3 leaves a ext3 sync-snap in client1's tree in S. This snap is used the next time ext3 is synced (step 7)

This sync-snap is deleted in step 4 without --no-rollback.

Step 7 fails as the ext3 sync-snap is gone.


There may be a better way of achieving what I'm doing here.

If not, it would be really good to have the -F flag optional in syncoid.

**Comment by jimsalterjrs** [comment]
@prewashed , I don't understand your concern. This isn't about sync snaps, which are entirely orthogonal to whether we explicitly roll back or use -F.

The default behavior of syncoid is still to use sync snaps.

**Comment by prewashed** [comment]
@jimsalterjrs, the -F deletes any subsequent sync snapshots of that target, that are made by syncoids of that target 

I thought these deleted sync snaps were required for the external sync to work. It turns out they aren't.

I am syncing the original target to multiple external disks. Each of these syncs sets a sync snap --identifier above and within the original target.

The external sync snaps that are in the original target get deleted by -F when the original data set is updated, as they are newer.

syncoid after completing the sync:

````
could not find any snapshots to destroy; check snapshot names.
WARNING:   zfs destroy 'ext1/S/backup/client1'@syncoid_ext3_hostname_2022-04-11:11:54:44-GMT10:00 failed: 256 at /usr/local/sbin/syncoid line 1363.
````

I did some testing and syncoid finds the latest matching snaps in the original target and the external disk and does its work.

It would be neater if the -F was optional, but it seems to work with -F and I can ignore the warning.

Thanks for the work, this is an excellent tool!

**Comment by jimsalterjrs** [comment]
Any successful replication destroys any snapshots on the target more recent 
than the newest snapshot in the incoming stream, period, whether you use 
-F, roll back manually, or both. There's literally no way around that. 
OpenZFS does not support a "multiverse" snapshot model; all snapshots on 
any given dataset or zvol are strictly linear (meaning you cannot add a new 
snapshot without wiping everything newer, whether it's a snapshot or simply 
data in the live filesystem).

**Comment by redmop** [comment]
If you want to branch from a snapshot, you need to clone, and start taking snapshots of it.

Note that cloning will mean that the snapshot you cloned from cannot deleted while the clone exists.

**Comment by prewashed** [comment]
I think for most cases, you're completely correct. If the target is modified, the receive will fail. Blowing anything new away with -F is a conservative default.

However if you are careful, you can make a later snapshot of that target if it is unchanged - if it was recieved unmounted and never mounted rw. This snapshot is effectively the same as the most recent snapshot on that target. Later you can receive over that, the foreign snapshot just sits there. This isn't technically branching, it is setting a placeholder in the target that is specific to that particular external backup. Like a bookmark in a way. I haven't explored using bookmarks for this.

syncoid from the client with "--no-rollback --recvoptions=u --identifier=client1" and from the target to the external backup with "--no-rollback --recvoptions=u --identifier=extN"

If you have an archival external disk, it now has an explicit snapshot on the target. that is impervious to deletion of other snapshots on the target.

With -F, sending to external still works as long as there are common snapshots. However, there is no guarantee that after some time and some deletion of snaps on the target that an older external disk will have a common snapshot. If that happens, the next send to the external will have to be a full send, not incremental. Not the end of the world.

e.g. using the old version of syncoid

the new "data" filesystem is successfully transferred to ext1, foreign snapshots interspersed with native ones.

````
for i in client1 S ext1 ext2; do truncate -s 1G $i; zpool create $i "$PWD"/$i; done

for i in ext1 ext2; do zfs create "$i/S"; done
zfs create S/backup

# backup the client
./syncoid -r --no-rollback --recvoptions=u --identifier=client1 client1 S/backup/client1

# backup to both "external disks"
./syncoid -r --identifier=ext1 --recvoptions='u' S/backup ext1/S/backup
./syncoid -r --identifier=ext2 --recvoptions='u' S/backup ext2/S/backup

# alter the client, backup
zfs create client1/data
./syncoid -r --recvoptions=u --identifier=client1 client1 S/backup/client1

# backup to ext1
./syncoid -r --identifier=ext1 --recvoptions='u' S/backup ext1/S/backup

# result:
========= client1 ==========
client1@syncoid_client1_hostname_2022-04-12:12:44:53-GMT10:00
client1/data@syncoid_client1_hostname_2022-04-12:12:44:53-GMT10:00
========= S ==========
S/backup@syncoid_ext2_hostname_2022-04-12:12:44:42-GMT10:00
S/backup@syncoid_ext1_hostname_2022-04-12:12:45:03-GMT10:00
S/backup/client1@syncoid_ext2_hostname_2022-04-12:12:44:42-GMT10:00
S/backup/client1@syncoid_client1_hostname_2022-04-12:12:44:53-GMT10:00
S/backup/client1@syncoid_ext1_hostname_2022-04-12:12:45:03-GMT10:00
S/backup/client1/data@syncoid_client1_hostname_2022-04-12:12:44:53-GMT10:00
S/backup/client1/data@syncoid_ext1_hostname_2022-04-12:12:45:03-GMT10:00
========= ext1 ==========
ext1/S/backup@syncoid_ext2_hostname_2022-04-12:12:44:42-GMT10:00
ext1/S/backup@syncoid_ext1_hostname_2022-04-12:12:45:03-GMT10:00
ext1/S/backup/client1@syncoid_client1_hostname_2022-04-12:12:44:28-GMT10:00
ext1/S/backup/client1@syncoid_ext2_hostname_2022-04-12:12:44:42-GMT10:00
ext1/S/backup/client1@syncoid_client1_hostname_2022-04-12:12:44:53-GMT10:00
ext1/S/backup/client1@syncoid_ext1_hostname_2022-04-12:12:45:03-GMT10:00
ext1/S/backup/client1/data@syncoid_client1_hostname_2022-04-12:12:44:53-GMT10:00
ext1/S/backup/client1/data@syncoid_ext1_hostname_2022-04-12:12:45:03-GMT10:00
========= ext2 ==========
ext2/S/backup@syncoid_ext1_hostname_2022-04-12:12:44:36-GMT10:00
ext2/S/backup@syncoid_ext2_hostname_2022-04-12:12:44:42-GMT10:00
ext2/S/backup/client1@syncoid_client1_hostname_2022-04-12:12:44:28-GMT10:00
ext2/S/backup/client1@syncoid_ext1_hostname_2022-04-12:12:44:36-GMT10:00
ext2/S/backup/client1@syncoid_ext2_hostname_2022-04-12:12:44:42-GMT10:00
